delivery_postlogistics/models/stock_picking.py

```diff
@@ -0,0 +1,230 @@
+        if "default_type" in context_attachment:
+            del context_attachment["default_type"]
+        return (
+            self.env["postlogistics.shipping.label"]
+            .with_context(context_attachment)
+            .create(data)
+        )
+
+    def _set_a_default_package(self):
+        """ Pickings using this module must have a package
+            If not this method put it one silently
+        """
+        for picking in self:
+            move_lines = picking.move_line_ids.filtered(
+                lambda s: not (s.package_id or s.result_package_id)
+            )
+            if move_lines:
+                default_packaging = (
+                    picking.carrier_id.postlogistics_default_packaging_id
+                )
+                package = self.env["stock.quant.package"].create(
+                    {
+                        "packaging_id": default_packaging
+                        and default_packaging.id
+                        or False
+                    }
+                )
+                move_lines.write({"result_package_id": package.id})
+
+    def postlogistics_cod_amount(self):
+        """ Return the Postlogistic Cash on Delivery amount of a picking
+
+        If the picking delivers the whole sales order, we use the total
+        amount of the sales order.
+
+        Otherwise, we don't know the value of each picking so we raise
+        an error.  The user has to create packages with the cash on
+        delivery price on each package.
+        """
+        self.ensure_one()
+        order = self.sale_id
+        if not order:
+            return 0.0
+        if len(order) > 1:
+            raise exceptions.Warning(
+                _(
+                    "The cash on delivery amount must be manually specified "
+                    "on the packages when a package contains products "
+                    "from different sales orders."
+                )
+            )
+        order_moves = order.mapped("order_line.procurement_ids.move_ids")
+        picking_moves = self.move_lines
+        # check if the package delivers the whole sales order
+        if order_moves != picking_moves:
+            raise exceptions.Warning(
+                _(
+                    "The cash on delivery amount must be manually specified "
+                    "on the packages when a sales order is delivered "
+                    "in several delivery orders."
+                )
+            )
+        return order.amount_total
+
+    def write_tracking_number_label(self, label_result, packages):
+        """
+        If there are no pack defined, write tracking_number on picking
+        otherwise, write it on parcel_tracking field of each pack.
+        Note we can receive multiple labels for a same package
+        """
+
+        def info_from_label(label):
+            tracking_number = label["tracking_number"]
+            return {
+                "file": base64.b64decode(label["binary"]),
```

Comment:
After base64 decoding, for **ZPL labels only** (not for PDF), you also have to call ".decode('cp437')"
Then the label is missing the enabling of UTF-8 on the printer. So you have to append after "^XA" the code "^CI28"

Reply:
For the context, when the city is "Zürich", the received data is encoded in a way that it cannot be printed

Reply:
I receive a file, if you requested a ZPL -> Call ".decode('cp437')". Then
"^XA" -> "^XA^CI28"

Reply:
Pay attention the content of base64.b64decode(label["binary"]) is still in base64. So it requires a second decoding before decoding with the right cp

Reply:
@jbaudoux, yes, I see it. do you think it was done on purpose ? I actually need to decode twice to get the content

Reply:
We are making a PR with the fix

Reply:
https://github.com/trobz/delivery-carrier/pull/2

Reply:
@santostelmo thanks for your PR, merged